### DISCUSSION TITLE
Internal self-contained migration?

### DISCUSSION THREAD

**Discussion by AlbertXiaoPeng** [casdoor/casdoor#2734] [Ideas]
Migration at the database level is not very well done and may require an internal system import and export for internal backups and restores.

**Comment by hsluoyz** [comment]
The open source version only makes best efforts but doesn't guarantee the DB can be fully auto migrated. Sometimes, you need to manually update some data to make it work.

**Reply by luv2code** [reply]
I think it would be helpful to provide some guidance on how to do this. You could just document how you guys handle it on your own hosted solution. I would like to see at least a note in the change log that indicates when a version has db schema changes.

At the moment, we upgrade and pray nothing is broken. Then if it is, how do we find out which db schema changes need to be made? My only guess is to diff the tags between what was deployed previously and what was just deployed, and look at the ORM code to try to guess the schema changes. That is overly cumbersome and error prone.
